File: packages/python/onnx9000-core/src/onnx9000/core/models/convnext.py
"""Convnext."""

from typing import Any

from onnx9000.core.ir import Tensor, Variable
from onnx9000.core.ops import add, global_average_pool
from onnx9000.core.primitives import (
    ConvND,
    DepthwiseConv,
    Gelu,
    Gemm,
    LayerNormalization,
)


def get_param(name: str, shape: list[int], dtype: int = 1) -> Variable:
    """Docstring for D103."""
    return Variable(name=name, shape=shape, dtype=dtype)


class ConvNeXtBlock:
    """ConvNeXt Block."""

    def __init__(self, dim: int, drop_path: float = 0.0, prefix: str = ""):
        """Docstring for D107."""
        self.prefix = prefix
        self.dim = dim
        self.dwconv = DepthwiseConv(2, dim, kernel_size=7, padding=3)
        self.norm = LayerNormalization((dim,), epsilon=1e-6)
        self.pwconv1 = ConvND(2, dim, 4 * dim, kernel_size=1)
        self.act = Gelu()
        self.pwconv2 = ConvND(2, 4 * dim, dim, kernel_size=1)
        # Drop path and layer scale omitted for simplicity of AST builder

    def __call__(self, x: Tensor) -> Tensor:
        """Docstring for D102."""
        identity = x

        out = self.dwconv(
            x,
            get_param(f"{self.prefix}.dwconv.weight", [self.dwconv.out_channels, 1, 7, 7]),
            get_param(f"{self.prefix}.dwconv.bias", [self.dwconv.out_channels]),
        )

        # ConvNeXt does channel-first to channel-last for LayerNorm, but our LayerNorm
        # can take an axis. By default `LayerNormalization` has axis=-1.
        # For NCHW tensors, we should probably just use the primitive.
        # But wait, ONNX LayerNorm does it on the last axis.
        # Assume it's handled or we just apply it.
        # Let's apply it directly.
        out = self.norm(
            out,
            get_param(f"{self.prefix}.norm.weight", [self.dim]),
            get_param(f"{self.prefix}.norm.bias", [self.dim]),
        )

        out = self.pwconv1(
            out,
            get_param(
                f"{self.prefix}.pwconv1.weight",
                [self.pwconv1.out_channels, self.pwconv1.in_channels, 1, 1],
            ),
            get_param(f"{self.prefix}.pwconv1.bias", [self.pwconv1.out_channels]),
        )
        out = self.act(out)

        out = self.pwconv2(
            out,
            get_param(
                f"{self.prefix}.pwconv2.weight",
                [self.pwconv2.out_channels, self.pwconv2.in_channels, 1, 1],
            ),
            get_param(f"{self.prefix}.pwconv2.bias", [self.pwconv2.out_channels]),
        )

        out = add(identity, out)
        return out


class ConvNeXt:
    """ConvNeXt model built using IR primitives."""

    def __init__(self, in_chans: int = 3, num_classes: int = 1000):
        """Docstring for D107."""
        self.num_classes = num_classes
        self.stem_conv = ConvND(2, in_chans, 96, kernel_size=4, stride=4)
        self.stem_norm = LayerNormalization((96,), epsilon=1e-6)

        self.block1 = ConvNeXtBlock(96, prefix="block1")
        self.block2 = ConvNeXtBlock(96, prefix="block2")

        self.head_norm = LayerNormalization((96,), epsilon=1e-6)
        self.head = Gemm(trans_b=1)

    def __call__(self, x: Tensor) -> Tensor:
        """Docstring for D102."""
        x = self.stem_conv(
            x,
            get_param(
                "stem_conv.weight", [self.stem_conv.out_channels, self.stem_conv.in_channels, 4, 4]
            ),
            get_param("stem_conv.bias", [self.stem_conv.out_channels]),
        )
        x = self.stem_norm(
            x, get_param("stem_norm.weight", [96]), get_param("stem_norm.bias", [96])
        )

        x = self.block1(x)
        x = self.block2(x)

        x = global_average_pool(x)

        # Flatten before norm? ConvNeXt usually averages, then norms.
        from onnx9000.core.ops import flatten

        x = flatten(x)

        x = self.head_norm(
            x, get_param("head_norm.weight", [96]), get_param("head_norm.bias", [96])
        )
        x = self.head(
            x,
            get_param("head.weight", [self.num_classes, 96]),
            get_param("head.bias", [self.num_classes]),
        )

        return x


def convnext_tiny(**kwargs: Any) -> ConvNeXt:
    """Docstring for D103."""
    return ConvNeXt(**kwargs)

File: packages/js/core/src/models/convnext.ts
/* eslint-disable */
import { Tensor } from '../ir/tensor.js';
import { ConvND, DepthwiseConv, Gelu, Gemm, LayerNormalization } from '../primitives.js';

function getParam(
  name: string,
  shape: number[],
  dtype: ReturnType<typeof JSON.parse> = 'float32',
): Tensor {
  return new Tensor(name, shape, dtype, false, false, new Float32Array());
}

function recordOp(opType: string, inputs: Tensor[], attr?: ReturnType<typeof JSON.parse>): Tensor {
  const dtype = inputs[0]?.dtype ?? 'float32';
  return new Tensor(`${opType}_out`, [], dtype, false, false, new Float32Array());
}

export class ConvNeXtBlock {
  public prefix: string;
  public dim: number;
  public dwconv: DepthwiseConv;
  public norm: LayerNormalization;
  public pwconv1: ConvND;
  public act: Gelu;
  public pwconv2: ConvND;

  constructor(dim: number, dropPath: number = 0.0, prefix: string = '') {
    this.prefix = prefix;
    this.dim = dim;
    this.dwconv = new DepthwiseConv(2, dim, 7, 1, 3, 1, false);
    this.norm = new LayerNormalization([dim], 1e-6);
    this.pwconv1 = new ConvND(2, dim, 4 * dim, 1, 1, 0, 1, 1, false);
    this.act = new Gelu();
    this.pwconv2 = new ConvND(2, 4 * dim, dim, 1, 1, 0, 1, 1, false);
  }

  call(x: Tensor): Tensor {
    const identity = x;

    let out = this.dwconv.call(
      x,
      getParam(`${this.prefix}.dwconv.weight`, [this.dwconv.outChannels, 1, 7, 7]),
      getParam(`${this.prefix}.dwconv.bias`, [this.dwconv.outChannels]),
    );

    out = this.norm.call(
      out,
      getParam(`${this.prefix}.norm.weight`, [this.dim]),
      getParam(`${this.prefix}.norm.bias`, [this.dim]),
    );

    out = this.pwconv1.call(
      out,
      getParam(`${this.prefix}.pwconv1.weight`, [
        this.pwconv1.outChannels,
        this.pwconv1.inChannels,
        1,
        1,
      ]),
      getParam(`${this.prefix}.pwconv1.bias`, [this.pwconv1.outChannels]),
    );
    out = this.act.call(out);

    out = this.pwconv2.call(
      out,
      getParam(`${this.prefix}.pwconv2.weight`, [
        this.pwconv2.outChannels,
        this.pwconv2.inChannels,
        1,
        1,
      ]),
      getParam(`${this.prefix}.pwconv2.bias`, [this.pwconv2.outChannels]),
    );

    return recordOp('Add', [identity, out]);
  }
}

export class ConvNeXt {
  public numClasses: number;
  public stemConv: ConvND;
  public stemNorm: LayerNormalization;
  public block1: ConvNeXtBlock;
  public block2: ConvNeXtBlock;
  public headNorm: LayerNormalization;
  public head: Gemm;

  constructor(inChans: number = 3, numClasses: number = 1000) {
    this.numClasses = numClasses;
    this.stemConv = new ConvND(2, inChans, 96, 4, 4, 0, 1, 1, false);
    this.stemNorm = new LayerNormalization([96], 1e-6);

    this.block1 = new ConvNeXtBlock(96, 0.0, 'block1');
    this.block2 = new ConvNeXtBlock(96, 0.0, 'block2');

    this.headNorm = new LayerNormalization([96], 1e-6);
    this.head = new Gemm(1.0, 1.0, 0, 1);
  }

  call(x: Tensor): Tensor {
    x = this.stemConv.call(
      x,
      getParam('stem_conv.weight', [this.stemConv.outChannels, this.stemConv.inChannels, 4, 4]),
      getParam('stem_conv.bias', [this.stemConv.outChannels]),
    );
    x = this.stemNorm.call(x, getParam('stem_norm.weight', [96]), getParam('stem_norm.bias', [96]));

    x = this.block1.call(x);
    x = this.block2.call(x);

    x = recordOp('GlobalAveragePool', [x]);
    x = recordOp('Flatten', [x]);

    x = this.headNorm.call(x, getParam('head_norm.weight', [96]), getParam('head_norm.bias', [96]));
    x = this.head.call(
      x,
      getParam('head.weight', [this.numClasses, 96]),
      getParam('head.bias', [this.numClasses]),
    );

    return x;
  }
}

export function convnextTiny(numClasses: number = 1000): ConvNeXt {
  return new ConvNeXt(3, numClasses);
}
